### ISSUE TITLE
quality metrics

### ISSUE THREAD

**Issue by WorldForever1** [iwuqing/ArSSR#3]
Hi，
Thanks for the code and paper. It is interesting to train a model for Arbitrary SR.
First one,  I don't konw why the T1WIs that I download from HCP website, such as, 531536_3T_T1w_MPR1.nii.gz,  are different from the 531536.nii.gz  you provide. So，do you make some preprocessing for the original T1Ws MR images, such as, N4 bias correction and skull-stripping? 

Second one, when I try to reproduce this experiment, there are some questions about quality metrics, LPIPS, PSI,LPC-SI. How to use the slice-by-slice strategy to compute them?  Would mind sharing the code about this partment? 

Look forward to any replies from you!
Best regards.

**Comment by WorldForever1** [comment]
Should I download the preprocessed data？ such as,  100307_3T_Structural_preproc.zip/T1w/T1w_acpc_dc_restore_brain.nii.gz. Thanks!

**Comment by 1632325673** [comment]
Is your problem solved？  Can you tell me how to get the test set？

**Comment by sinAshish** [comment]
any solutions for the above questions?
for the pre-processing I can see that they have provided the data_bulid.py.

**Comment by iwuqing** [comment]
> Should I download the preprocessed data？ such as, 100307_3T_Structural_preproc.zip/T1w/T1w_acpc_dc_restore_brain.nii.gz. Thanks!
Yeah,  the HCP-1200 dataset we download is . We use a Python script (https://github.com/huawei-lin/HCP_Dataset_Download_Automatically_Script) to perform downloading.  To download the image modality we want (e.g., T1w in our paper), we modify the lines 49-51 this script  as blew:

Original in https://github.com/huawei-lin/HCP_Dataset_Download_Automatically_Script/blob/master/script.py
--------
`   keyList = bucket.objects.filter(Prefix = prefix + '/{}/MNINonLinear/Results/tfMRI'.format(subject_number))
    keyList = [key.key for key in keyList]
    keyList = [x for x in keyList if '_LR.nii.gz' in x or '_RL.nii.gz' in x or 'EVs' in x]
    
    if not os.path.exists(output_path):
        os.makedirs(output_path)
        
    totalNumber = len(keyList)
    trycnt = 0
    
    
    
    tempKeys = [x for x in keyList if '_LR.nii.gz' in x or '_RL.nii.gz' in x]`

Modified 
--------
`   keyList = bucket.objects.filter(Prefix = prefix + '/{}/MNINonLinear'.format(subject_number))
    keyList = [key.key for key in keyList]
    keyList = [x for x in keyList if 'T2w_restore_brain.nii.gz' in x]
    
    if not os.path.exists(output_path):
        os.makedirs(output_path)
        
    totalNumber = len(keyList)
    trycnt = 0
    
    tempKeys = [x for x in keyList if 'T2w_restore_brain.nii.gz' in x]`

**Comment by iwuqing** [comment]
> Second one, when I try to reproduce this experiment, there are some questions about quality metrics, LPIPS, PSI,LPC-SI. How to use the slice-by-slice strategy to compute them? Would mind sharing the code about this partment?

The three metrics are employed from their official implementation, their code links are:
LPIPS：https://github.com/richzhang/PerceptualSimilarity
PSI：https://github.com/feichtenhofer/PSI
LPC-SI：https://ece.uwaterloo.ca/~z70wang/research/lpcsi/

Moreover, the slice-by-slice strategy includes three steps:
1. Given a 3D volume, we extract its 2D MR slices from three orthogonal directions (axial, sagittal, and coronal directions). 
2. We compute the scores for each 2D MR slice.
3. We average the scores of all the 2D MR slices to calculate the final scores.

The following code is to compute LIPIPS using the slice-by-slice strategy:
```python
import SimpleITK as sitk
import numpy as np
import torch
import lpips

def compute_lpips(gt_slice, recon_slice):
    h, w = gt_slice.shape
    gt_image = np.zeros((1, 3, h, w))
    recon_image = np.zeros((1, 3, h, w))
    for c in range(3):
        gt_image[:, c, :, :] = gt_slice
        recon_image[:, c, :, :] = recon_slice
    lpips_value = lpips_alex(torch.tensor(gt_image).to(torch.float32), torch.tensor(recon_image).to(torch.float32))
    return lpips_value

lpips_alex = lpips.LPIPS(net = 'alex')

gt_volume_path = ''
recon_volume_path = ''

gt_volume = sitk.GetArrayFromImage(sitk.ReadImage(gt_volume_path))
recon_volume = sitk.GetArrayFromImage(sitk.ReadImage(recon_volume_path))

# In any volume, the center region of volume includes more image information compared with mergin region.
# Therefore, we extract 10 2D slices in the center from three directions, resectively.
# Here the volumes are 264*264*264 size
i_strart, i_end = 127, 137

lpips_value = 0.

for i in range(i_strart, i_end):
    # direction 1
    gt_slice = gt_volume[i, :, :]
    recon_slice = recon_volume[i, :, :]
    lpips_value += compute_lpips(gt_slice, recon_slice)
    # direction 2
    gt_slice = gt_volume[:, i, :]
    recon_slice = recon_volume[:, i, :]
    lpips_value += compute_lpips(gt_slice, recon_slice)
    # direction 3
    gt_slice = gt_volume[:, :, i]
    recon_slice = recon_volume[:, :, i]
    lpips_value += compute_lpips(gt_slice, recon_slice)

lpips_value = lpips_value / 30
print('LPIPS:', lpips_value)
```

**Comment by WorldForever1** [comment]
Thanks, I will have a try!
